src/Mvc/Mvc.ViewFeatures/src/HtmlHelperComponentExtensions.cs

```diff
@@ -37,30 +39,85 @@ public static class HtmlHelperComponentExtensions
         /// <param name="htmlHelper">The <see cref="IHtmlHelper"/>.</param>
         /// <param name="parameters">An <see cref="object"/> containing the parameters to pass
         /// to the component.</param>
+        /// <param name="renderMode">The <see cref="RenderMode"/> for the component.</param>
         /// <returns>The HTML produced by the rendered <typeparamref name="TComponent"/>.</returns>
         public static async Task<IHtmlContent> RenderComponentAsync<TComponent>(
             this IHtmlHelper htmlHelper,
+            RenderMode renderMode,
             object parameters) where TComponent : IComponent
         {
             if (htmlHelper == null)
             {
                 throw new ArgumentNullException(nameof(htmlHelper));
             }
 
-            var httpContext = htmlHelper.ViewContext.HttpContext;
-            var serviceProvider = httpContext.RequestServices;
-            var prerenderer = serviceProvider.GetRequiredService<StaticComponentRenderer>();
+            var context = htmlHelper.ViewContext.HttpContext;
+            return renderMode switch
+            {
+                RenderMode.Server => NonPrerenderedBlazorComponent(context, typeof(TComponent), GetPrameterCollection(parameters)),
+                RenderMode.ServerPrerendered => await PrerenderedBlazorComponentAsync(context, typeof(TComponent), GetPrameterCollection(parameters)),
+                RenderMode.Static => await StaticComponentAsync(context, typeof(TComponent), GetPrameterCollection(parameters)),
+                _ => throw new ArgumentException("Invalid render mode", nameof(renderMode)),
+            };
+        }
 
-            var parametersCollection = parameters == null ?
+        private static ParameterView GetPrameterCollection(object parameters) => parameters == null ?
                 ParameterView.Empty :
                 ParameterView.FromDictionary(HtmlHelper.ObjectToDictionary(parameters));
 
+        private static async Task<IHtmlContent> StaticComponentAsync(HttpContext context, Type type, ParameterView parametersCollection)
+        {
+            var serviceProvider = context.RequestServices;
+            var prerenderer = serviceProvider.GetRequiredService<StaticComponentRenderer>();
+
+
             var result = await prerenderer.PrerenderComponentAsync(
                 parametersCollection,
-                httpContext,
-                typeof(TComponent));
+                context,
+                type);
 
             return new ComponentHtmlContent(result);
         }
+
+        private static async Task<IHtmlContent> PrerenderedBlazorComponentAsync(HttpContext context, Type type, ParameterView parametersCollection)
+        {
+            if (parametersCollection.GetEnumerator().MoveNext())
```

Comment:
You could simplify this by passing in the dictionary to these methods and creating the parameter view inside

Reply:
I want to avoid the allocation when parameters == null, checking on parameterView is cheap thanks to the struct enum optimization